esp/services/ws_smc/ws_smcService.hpp

```diff
@@ -154,15 +157,69 @@ class CActivityInfo : public CInterface, implements IInterface
     inline IArrayOf<IEspActiveWorkunit>& queryActiveWUs() { return aws; };
     inline IArrayOf<IEspServerJobQueue>& queryServerJobQueues() { return serverJobQueues; };
     inline IArrayOf<IEspDFUJob>& queryDFURecoveryJobs() { return DFURecoveryJobs; };
+    inline StringBuffer& queryTimeCached(StringBuffer& str) { return timeCached.getString(str, true); }
+};
+
+class CWsSMCEx;
+
+class CActivityInfoReader : public Thread
+{
+    bool stopping = false;
+    bool detached = false;
+    unsigned autoRebuildMillSeconds = 1000*DEFAULTACTIVITYINFOCACHEAUTOREBUILDSECOND;
+    unsigned forceRebuildSeconds = DEFAULTACTIVITYINFOCACHEFORCEBUILDSECOND;
+    Owned<CActivityInfo> activityInfoCache;
+    Semaphore sem;
+    CriticalSection crit;
+public:
+    CActivityInfoReader(unsigned _autoRebuildMillSeconds, unsigned _forceRebuildSeconds)
+        : autoRebuildMillSeconds(_autoRebuildMillSeconds), forceRebuildSeconds(_forceRebuildSeconds) {};
+
+    ~CActivityInfoReader()
+    {
+        stopping = true;
+        sem.signal();
+        join();
+    }
+
+    virtual int run();
+    CActivityInfo* getActivityInfo()
+    {
+        while (!stopping)
+        {
+            {
+                CriticalBlock b(crit);
+                if (activityInfoCache)
```

Comment:
see other comment below (after line 199)
If run has been changed to block if no existing activityInfoCache entry, then you can remove condition here (just assert), you can also remove the sleep and loop.

Reply:
Will do.